File: backend/app/api/routes/agents.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query

from app import crud
from app.api.deps import CurrentUser, SessionDep, get_current_user
from app.models import (
    Agent,
    AgentCreate,
    AgentDraftUpdate,
    AgentPublic,
    AgentRollbackRequest,
    AgentsPublic,
    AgentUpdate,
    AgentVersionDiff,
    AgentVersionPublic,
    AgentVersionsPublic,
    Message,
    PublishRequest,
)
from app.services.diff import compute_agent_version_diff

router = APIRouter(
    prefix="/agents", tags=["agents"], dependencies=[Depends(get_current_user)]
)


@router.post("/", response_model=AgentPublic)
def create_agent(
    session: SessionDep,
    current_user: CurrentUser,
    agent_in: AgentCreate,
) -> Agent:
    return crud.create_agent(
        session=session, agent_in=agent_in, created_by=current_user.id
    )


@router.get("/", response_model=AgentsPublic)
def list_agents(
    session: SessionDep,
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1, le=1000),
) -> AgentsPublic:
    items, count = crud.list_agents(session=session, skip=skip, limit=limit)
    return AgentsPublic(data=items, count=count)  # type: ignore[arg-type]


@router.get("/{agent_id}/versions/diff", response_model=AgentVersionDiff)
def diff_agent_versions(
    session: SessionDep,
    agent_id: uuid.UUID,
    from_version: int = Query(ge=1, description="Earlier version number"),
    to_version: int = Query(ge=1, description="Later version number"),
) -> AgentVersionDiff:
    if from_version == to_version:
        raise HTTPException(
            status_code=400,
            detail="from_version and to_version must be different",
        )
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    row_from = crud.get_agent_version(
        session=session, agent_id=agent_id, version=from_version
    )
    row_to = crud.get_agent_version(
        session=session, agent_id=agent_id, version=to_version
    )
    if row_from is None or row_to is None:
        raise HTTPException(status_code=404, detail="One or both versions not found")
    return compute_agent_version_diff(
        from_version_num=from_version,
        to_version_num=to_version,
        old=row_from,
        new=row_to,
    )


@router.get("/{agent_id}/versions/{version}", response_model=AgentVersionPublic)
def read_agent_version(
    session: SessionDep,
    agent_id: uuid.UUID,
    version: int,
) -> AgentVersionPublic:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    row = crud.get_agent_version(session=session, agent_id=agent_id, version=version)
    if row is None:
        raise HTTPException(status_code=404, detail="Version not found")
    return AgentVersionPublic.model_validate(row)


@router.get("/{agent_id}/versions", response_model=AgentVersionsPublic)
def list_agent_versions(
    session: SessionDep,
    agent_id: uuid.UUID,
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1, le=1000),
) -> AgentVersionsPublic:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    items, count = crud.list_agent_versions(
        session=session, agent_id=agent_id, skip=skip, limit=limit
    )
    return AgentVersionsPublic(
        data=[AgentVersionPublic.model_validate(v) for v in items],
        count=count,
    )


@router.post("/{agent_id}/rollback", response_model=AgentVersionPublic)
def rollback_agent(
    session: SessionDep,
    current_user: CurrentUser,
    agent_id: uuid.UUID,
    body: AgentRollbackRequest,
) -> AgentVersionPublic:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    try:
        _agent, new_row = crud.rollback_agent_version(
            session=session,
            db_agent=agent,
            target_version=body.version,
            change_description=body.change_description,
            created_by=current_user.id,
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return AgentVersionPublic.model_validate(new_row)


# ── Draft / Publish endpoints ────────────────────────────────────────


@router.put("/{agent_id}/draft", response_model=AgentVersionPublic)
def upsert_draft(
    session: SessionDep,
    current_user: CurrentUser,
    agent_id: uuid.UUID,
    body: AgentDraftUpdate,
) -> AgentVersionPublic:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    draft_data = body.model_dump(exclude_unset=True)
    if not draft_data:
        raise HTTPException(status_code=422, detail="No fields provided")
    try:
        draft = crud.create_or_update_agent_draft(
            session=session,
            agent=agent,
            draft_data=draft_data,
            created_by=current_user.id,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e
    return AgentVersionPublic.model_validate(draft)


@router.get("/{agent_id}/draft", response_model=AgentVersionPublic)
def get_draft(
    session: SessionDep,
    agent_id: uuid.UUID,
) -> AgentVersionPublic:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    draft = crud.get_agent_draft(session=session, agent_id=agent_id)
    if draft is None:
        raise HTTPException(status_code=404, detail="No draft found")
    return AgentVersionPublic.model_validate(draft)


@router.post("/{agent_id}/publish", response_model=AgentVersionPublic)
def publish_draft(
    session: SessionDep,
    current_user: CurrentUser,
    agent_id: uuid.UUID,
    body: PublishRequest,
) -> AgentVersionPublic:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    try:
        published = crud.publish_agent_draft(
            session=session,
            agent=agent,
            change_description=body.change_description,
            created_by=current_user.id,
        )
    except ValueError as e:
        detail = str(e)
        if detail == "No draft to publish":
            raise HTTPException(status_code=409, detail=detail) from e
        raise HTTPException(status_code=422, detail=detail) from e
    return AgentVersionPublic.model_validate(published)


@router.delete("/{agent_id}/draft", status_code=204)
def discard_draft(
    session: SessionDep,
    agent_id: uuid.UUID,
) -> None:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    crud.discard_agent_draft(session=session, agent=agent)


# ── Agent CRUD endpoints ─────────────────────────────────────────────


@router.get("/{agent_id}", response_model=AgentPublic)
def get_agent(session: SessionDep, agent_id: uuid.UUID) -> Agent:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return agent


@router.patch("/{agent_id}", response_model=AgentPublic)
def update_agent(
    session: SessionDep,
    current_user: CurrentUser,
    agent_id: uuid.UUID,
    agent_in: AgentUpdate,
) -> Agent:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    try:
        return crud.update_agent(
            session=session,
            db_agent=agent,
            agent_in=agent_in,
            created_by=current_user.id,
        )
    except ValueError as e:
        detail = str(e)
        if detail == "Agent not found":
            raise HTTPException(status_code=404, detail=detail) from e
        raise HTTPException(status_code=422, detail=detail) from e


@router.delete("/{agent_id}", response_model=Message)
def delete_agent(session: SessionDep, agent_id: uuid.UUID) -> Message:
    agent = crud.get_agent(session=session, agent_id=agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    crud.delete_agent(session=session, db_agent=agent)
    return Message(message="Agent deleted successfully")

File: frontend/apps/web/src/actions/agents.ts
'use server';

import { AgentsService } from '@/client/sdk.gen';
import { isSuccessApiResult } from '@/utils/api';

import type {
  AgentCreate,
  AgentDraftUpdate,
  AgentPublic,
  AgentRollbackRequest,
  AgentUpdate,
  AgentVersionPublic,
  AgentVersionsPublic,
  PublishRequest,
} from '@/client/types.gen';
import type { ApiResult } from '@/types/api';

export type AgentRow = AgentPublic;

export type AgentFilters = {
  name?: string;
  page?: number;
  pageSize?: number;
};

export type GetAgentsResult = {
  rows: AgentRow[];
  totalCount: number;
};

export const getAgents = async (filters: AgentFilters = {}): Promise<GetAgentsResult> => {
  const page = filters.page ?? 1;
  const pageSize = filters.pageSize ?? 10;
  const skip = (page - 1) * pageSize;

  const apiResponse = await AgentsService.listAgents({
    query: { skip, limit: pageSize },
  });

  const { response: _, ...result } = apiResponse;

  if (!isSuccessApiResult(result)) {
    return { rows: [], totalCount: 0 };
  }

  return {
    rows: result.data.data,
    totalCount: result.data.count,
  };
};

export const deleteAgent = async (agentId: string): Promise<{ success: boolean }> => {
  const apiResponse = await AgentsService.deleteAgent({
    path: { agent_id: agentId },
  });

  const { response: _, ...result } = apiResponse;

  return { success: isSuccessApiResult(result) };
};

export const createAgent = async (body: AgentCreate): Promise<ApiResult<AgentPublic>> => {
  const apiResponse = await AgentsService.createAgent({ body });
  const { response: _, ...result } = apiResponse;
  return result;
};

export const createDraftAgent = async (
  name: string = 'Untitled Agent'
): Promise<ApiResult<AgentPublic>> => {
  const { client } = await import('@/client/client.gen');
  const apiResponse = await client.post({
    url: '/api/v1/agents/draft',
    body: { name },
    headers: { 'Content-Type': 'application/json' },
  });
  const { response: _, ...result } = apiResponse;
  return result as ApiResult<AgentPublic>;
};

export const getAgent = async (agentId: string): Promise<ApiResult<AgentPublic>> => {
  const apiResponse = await AgentsService.getAgent({
    path: { agent_id: agentId },
  });
  const { response: _, ...result } = apiResponse;
  return result;
};

export const updateAgent = async (
  agentId: string,
  body: AgentUpdate
): Promise<ApiResult<AgentPublic>> => {
  const apiResponse = await AgentsService.updateAgent({
    path: { agent_id: agentId },
    body,
  });
  const { response: _, ...result } = apiResponse;
  return result;
};

export const getAgentVersions = async (
  agentId: string,
  skip: number = 0,
  limit: number = 100
): Promise<ApiResult<AgentVersionsPublic>> => {
  const apiResponse = await AgentsService.listAgentVersions({
    path: { agent_id: agentId },
    query: { skip, limit },
  });
  const { response: _, ...result } = apiResponse;
  return result as ApiResult<AgentVersionsPublic>;
};

export const getAgentVersion = async (
  agentId: string,
  version: number
): Promise<ApiResult<AgentVersionPublic>> => {
  const apiResponse = await AgentsService.readAgentVersion({
    path: { agent_id: agentId, version },
  });
  const { response: _, ...result } = apiResponse;
  return result as ApiResult<AgentVersionPublic>;
};

export const getAgentDraft = async (
  agentId: string
): Promise<ApiResult<AgentVersionPublic>> => {
  const apiResponse = await AgentsService.getDraft({
    path: { agent_id: agentId },
  });
  const { response: _, ...result } = apiResponse;
  return result as ApiResult<AgentVersionPublic>;
};

export const upsertAgentDraft = async (
  agentId: string,
  body: AgentDraftUpdate
): Promise<ApiResult<AgentVersionPublic>> => {
  const apiResponse = await AgentsService.upsertDraft({
    path: { agent_id: agentId },
    body,
  });
  const { response: _, ...result } = apiResponse;
  return result as ApiResult<AgentVersionPublic>;
};

export const publishAgentDraft = async (
  agentId: string,
  body: PublishRequest
): Promise<ApiResult<AgentVersionPublic>> => {
  const apiResponse = await AgentsService.publishDraft({
    path: { agent_id: agentId },
    body,
  });
  const { response: _, ...result } = apiResponse;
  return result as ApiResult<AgentVersionPublic>;
};

export const rollbackAgentVersion = async (
  agentId: string,
  body: AgentRollbackRequest
): Promise<ApiResult<AgentVersionPublic>> => {
  const apiResponse = await AgentsService.rollbackAgent({
    path: { agent_id: agentId },
    body,
  });
  const { response: _, ...result } = apiResponse;
  return result as ApiResult<AgentVersionPublic>;
};
